### PR TITLE
TG-1212 Bugfix/add generic type args to dependencies

### PR DESCRIPTION
**Pull request by mgudemann**

### CODE REVIEW

File: src/java_bytecode/java_types.cpp [hunk 1]
```diff
@@ -715,3 +715,52 @@ bool is_valid_java_array(const struct_typet &type)
...
+  // symbol type
+  else if(t.id() == ID_symbol)
+  {
+    const irep_idt class_name(
+      strip_java_namespace_prefix(to_symbol_type(t).get_identifier()));
+    refs.insert(class_name);
+  }
+}
+
+irep_idt strip_java_namespace_prefix(const irep_idt &to_strip)
```
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: This is surely a generic util rather than belonging under `java_types`

**Comment by smowton** [comment]
Guessing this move relates to the CI fail too? Ping me when resolved please

**Comment by mgudemann** [comment]
@smowton should be resolved, most CI targets have already succeeded now

**Review by smowton — APPROVED** [review]

**Comment by smowton** [comment]
Before merging could you verify if you have regression tests covering all the cases (bounded, unbounded, ...) in the new parse code?

**Comment by mgudemann** [comment]
@smowton the parsing code itself hasn't changed. The difference here is that we do an initial parse of the signatures, in order to find required dependencies of classes that are only referenced as generic type arguments. This only really concerns instantitated generic type arguments, so bounds shouldn't be relevant here.

File: src/java_bytecode/java_types.cpp [hunk 2]
```diff
@@ -715,3 +715,93 @@ bool is_valid_java_array(const struct_typet &type)
...
+/// \param signature: the string representation of the signature to analyze
+/// \param refs [out]: the set to insert the names of the found dependencies
+void get_dependencies_from_generic_parameters(
+  const std::string &signature,
+  std::set<irep_idt> &refs)
+{
+  try
+  {
+    // class signature with bounds
+    if(signature[0] == '<')
```
Comment (cesaro) [inline src/java_bytecode/java_types.cpp:773]: Is this the parsing that @smowton was referring to?

**Review by cesaro — APPROVED** [review]
Looks neat. Same reaction than Chris, would it be necessary to add more tests for the line I pointed out?

**Comment by mgudemann** [comment]
@cesaro What happens here is that the existing parser (nothing changed for that) is called for a signature of the form `<T::CBound>` or `<T:IBound>`. So in that case we'd add `CBound` and `IBound` as classes to load, too. This might be considered purely safety, as an implementing class for `IBound` or a derived class of `CBound` would be needed for an instantiation here and then the interface or the base class should be loaded in any case.

The main difference here is that class signatures of the above form are parsed via `java_generic_type_form_string`, while class signatures like `Enum<EnumVal>` are parsed via `java_type_from_string`. The former are class files where the generic type argument is already instantiated at compile-time.

**Comment by mgudemann** [comment]
@cesaro so in short, the cases we care about here are those that do not have bounds, but a concrete instantiation

**Comment by cesaro** [comment]
@mgudemann I see, thanks for the explanation

**Pull request merged**